controllers/upgrade_test.go

```diff
@@ -0,0 +1,527 @@
+		if actualValue != test.ExpectedValue {
+			t.Errorf("Test Description: %s \n expected value: %v, actual value: %v", test.TestDescription, test.ExpectedValue, actualValue)
+		}
+	}
+
+}
+
+func TestDesiredNodesReady(t *testing.T) {
+	var tests = []struct {
+		TestDescription string
+		Reconciler      *RollingUpgradeReconciler
+		RollingUpgrade  *v1alpha1.RollingUpgrade
+		AsgClient       *MockAutoscalingGroup
+		ClusterNodes    *corev1.NodeList
+		ExpectedValue   bool
+	}{
+		{
+			"Desired nodes are ready",
+			createRollingUpgradeReconciler(t),
+			createRollingUpgrade(),
+			createASGClient(),
+			createNodeList(),
+			true,
+		},
+		{
+			"Desired instances are not ready (desiredCount != inServiceCount)",
+			createRollingUpgradeReconciler(t),
+			createRollingUpgrade(),
+			func() *MockAutoscalingGroup {
+				newAsgClient := createASGClient()
+				newAsgClient.autoScalingGroups[0].DesiredCapacity = func(x int) *int64 { i := int64(x); return &i }(4)
+				return newAsgClient
+			}(),
+			createNodeList(),
+			false,
+		},
+		{
+			"None of the nodes are ready (desiredCount != readyCount)",
+			createRollingUpgradeReconciler(t),
+			createRollingUpgrade(),
+			createASGClient(),
+			func() *corev1.NodeList {
+				var nodeList = &corev1.NodeList{Items: []corev1.Node{}}
+				for i := 0; i < 3; i++ {
+					node := createNode()
+					node.Status.Conditions = []corev1.NodeCondition{
+						{Type: corev1.NodeReady, Status: corev1.ConditionFalse},
+					}
+					nodeList.Items = append(nodeList.Items, *node)
+				}
+				return nodeList
+			}(),
+			false,
+		},
+		{
+			"None of the instances are InService (desiredCount != inServiceCount)",
+			createRollingUpgradeReconciler(t),
+			createRollingUpgrade(),
+			func() *MockAutoscalingGroup {
+				newAsgClient := createASGClient()
+				newAsgClient.autoScalingGroups[0].Instances = []*autoscaling.Instance{
+					&autoscaling.Instance{InstanceId: aws.String("mock-instance-1"), LifecycleState: aws.String("Pending")},
+					&autoscaling.Instance{InstanceId: aws.String("mock-instance-2"), LifecycleState: aws.String("Terminating")},
+					&autoscaling.Instance{InstanceId: aws.String("mock-instance-3"), LifecycleState: aws.String("Terminating")},
+				}
+				return newAsgClient
+			}(),
+			createNodeList(),
+			false,
+		},
+	}
+	for _, test := range tests {
+		test.Reconciler.Cloud.ScalingGroups = test.AsgClient.autoScalingGroups
+		test.Reconciler.Cloud.ClusterNodes = test.ClusterNodes
+		test.Reconciler.Auth.AmazonClientSet.AsgClient = test.AsgClient
+
+		actualValue := test.Reconciler.DesiredNodesReady(test.RollingUpgrade)
+		if actualValue != test.ExpectedValue {
+			t.Errorf("Test Description: %s \n expected value: %v, actual value: %v", test.TestDescription, test.ExpectedValue, actualValue)
+		}
+	}
+}
+
+/********************************************************** HELPER FUNCTIONS **********************************************************/
```

Comment:
might be nicer to just move the below code into it's own file e.g. `helpers_test.go`